Build Fortran libraries

adamcpovey opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [X] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [X] Bumped the build number (if the version is unchanged)
* [X] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [X] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->
Closes #39 

- It's possible to keep using the tarballs rather than GitHub for the source, but you'd have to change to [the CMake file](https://support.hdfgroup.org/ftp/HDF/releases/HDF4.2.15/src/CMake-hdf-4.2.15.tar.gz) and that has a slightly different folder structure which will impact the Windows build.
- In my local environment, I could build successfully with `ctest` but could not do that with this repo for reasons I don't understand. Hence, I switched to a `cmake; make test` arrangement that I think does the same thing but might not.
- You will need to check with https://github.com/conda-forge/hdfeos2-feedstock that this doesn't break their builds as `h4fc` is not output. It's fine for me but, again, my local environment is weird.
- I haven't checked the Windows build as I don't have one of those computers. Sorry.
- I haven't done the re-rendering requested as the first step of that is "install conda smithy" and I super do not want to touch my root environment at this moment in time (being in the middle of something).

adamcpovey committed:
Fetch the source from GitHub rather than a tarball

adamcpovey committed:
Update max_file number test

adamcpovey committed:
Switch to cmake compilation on Linux

This permits building both shared and Fortran libraries, which is
disallowed by the configure scripts. It also changes the name of
libdf.* to libhdf.*; we add symlinks for the old names for backwards
compatability.

adamcpovey committed:
Changes to comments

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

matthiasdiener commented on recipe/meta.yaml:
```diff
@@ -5,35 +5,32 @@ package:
+    - cmake
     - bison       # [not win]
     - flex        # [not win]
-    - make        # [not win]
```
`make` is still necessary I think

adamcpovey replied:
I was assuming cmake brought that with it but you are correct that it doesn't.

matthiasdiener commented on recipe/meta.yaml:
```diff
@@ -5,35 +5,32 @@ package:
-    - patches/debian/hdf-4.2.10-aarch64.patch
-    - patches/debian/hdfi.h.patch
-    - patches/debian/spelling-errors.patch
+  - git_url: https://github.com/HDFGroup/hdf4
```
Out of curiosity, why are you using the git URL?

adamcpovey replied:
1. The easiest way to test compiling several versions of the code without keeping track of lots of sha256 checksums.
2. Personal preference for dynamic over static code bases.

matthiasdiener commented on recipe/build.sh:
```diff
@@ -14,38 +14,34 @@ if [[ $(uname) == Darwin ]]; then
+
+mkdir build
+cd build
+cmake -G "Unix Makefiles" \
```
I think this needs `${CMAKE_ARGS}` (https://conda-forge.org/blog/posts/2020-10-29-macos-arm64/)

adamcpovey replied:
I did wonder about that. On my system, they seem to get picked up automatically.

matthiasdiener commented on recipe/build.sh:
```diff
@@ -5,7 +5,7 @@ set -x
 cp $BUILD_PREFIX/share/gnuconfig/config.* .
 
-# The compiler flags interfere with th build and we need to overide them :-/
+# The compiler flags interfere with the build and we need to overide them :-/
```
I think we could try removing this code that unsets the flags.

adamcpovey replied:
Is the etiquette for me to do that or for you?

matthiasdiener replied:
I don't think there is a particular etiquette for this, it was just a suggestion that might help with compilation. If it works with these flags being unset, I think it is probably better to keep it like it is :-)

adamcpovey committed:
Things that were missing

- osx builds need CMAKE_ARGS to be included explicitly
- make is a dependency for unix builds
- libnetcdf links against libmfhdf.so.0

adamcpovey committed:
Merge branch 'main' into main

adamcpovey committed:
Switch back to static download

docker couldn't resolve the git tag so revert the source to the url/sha256 from the original repo
